### ISSUE TITLE
MoE forward operations not tape-compatible: gradients zero through routing path

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#1080]
## Bug Description

MixtureOfExpertsLayer's forward pass uses scalar NumOps loops for softmax routing and expert output combination. These operations are invisible to the GradientTape autodiff system, so gradients don't flow through the routing path — expert and router parameters get zero gradients even though the tape sees their weight tensors.

**Current status**: 15/17 MoE tests pass after sublayer registration fix (PR #1063). The remaining 2 failures are:
- `Training_ShouldChangeParameters` — parameters don't change (zero gradients)
- `GradientFlow_ShouldBeNonZeroAndFinite` — same root cause

## Root Cause

Three operations in `MixtureOfExpertsLayer.Forward()` use scalar loops instead of tape-tracked tensor operations:

### 1. ApplySoftmax (line ~1263)
```csharp
// Manual scalar loop — NOT tracked by tape
for (int b = 0; b < batchSize; b++)
{
    T maxVal = logits[b, 0];
    for (int j = 1; j < numExperts; j++)
        if (NumOps.ToDouble(logits[b, j]) > NumOps.ToDouble(maxVal))
            maxVal = logits[b, j];
    // ... exp, sum, divide per element
}
```

Should use: `Engine.Softmax(logits, axis: 1)` or equivalent tape-tracked operation.

### 2. CombineExpertOutputs (line ~1384)
```csharp
// Uses Engine.TensorBroadcastMultiply — may be tape-tracked
var weightedOutput = Engine.TensorBroadcastMultiply(expertOutput, weightColumn);
combined = Engine.TensorBroadcastAdd(combined, weightedOutput);
```

This might already be partially tape-compatible if `TensorBroadcastMultiply` records to the tape. Needs verification.

### 3. ApplyTopK (sparse routing)
```csharp
// Manual masking — NOT tracked by tape
for (int b = 0; b < batchSize; b++)
{
    // zero out non-top-K weights
}
```

Should use differentiable top-K with straight-through estimator or Gumbel-Softmax.

## Proposed Fix

Rewrite the three operations to use tape-tracked tensor operations:

1. **Softmax**: Use the engine's built-in softmax (or TensorOperations.Softmax from autodiff) 
2. **Combine**: Verify Engine broadcast ops are tape-tracked; if not, use TensorOperations equivalents
3. **TopK**: For training, use Gumbel-Softmax (Jang et al. 2017) for differentiable expert selection, falling back to hard TopK for inference

This matches PyTorch's approach where `torch.nn.functional.softmax` and tensor multiplication are autograd-compatible by default.

## References
- Shazeer et al. 2017 "Outrageously Large Neural Networks: The Sparsely-Gated Mixture-of-Experts Layer"
- Jang et al. 2017 "Categorical Reparameterization with Gumbel-Softmax" (for differentiable TopK)

**Comment by ooples** [comment]
## Audit Update (2026-04-06)

The MoE layer has been significantly updated since this issue was filed. Current state:

- **ApplySoftmax**: Fully vectorized via `Engine.TensorExp`, `Engine.ReduceSum`, `Engine.TensorBroadcastDivide` — NOT scalar NumOps loops
- **CombineExpertOutputs**: Uses `Engine.TensorBroadcastMultiply` and `Engine.TensorBroadcastAdd` — tape-tracked
- **ApplyTopK**: Uses `Engine.TensorTopK`, `Engine.TensorScatter`, `Engine.ReduceSum` — vectorized

The main forward path (routing → softmax → expert combination) now uses Engine ops throughout. The routing gradients should flow through the tape since all operations are tape-tracked.

**Remaining concerns:**
1. The TopK selection (`TensorTopK` + `TensorScatter`) may not be differentiable through the top-k index selection (hard routing). This is expected behavior per Switch Transformer — top-k is non-differentiable but load-balancing auxiliary loss provides the training signal for the router.
2. Sparse routing still uses scalar loops to check `IsExpertActive` per batch item (lines 593-600). This is metadata logic, not gradient-affecting.

**Recommendation:** This issue may be mostly resolved by recent Engine vectorization work. Consider closing if tests pass, or narrowing scope to just the TopK differentiability question.
